### PR TITLE
HTTP API

### PR DESCRIPTION
**Pull request by cbbayburt**
Implementation for the [HTTP API](https://github.com/uyuni-project/uyuni-rfc/blob/http-api/accepted/00000-http-api.md)

The PoC implements the boilerplate code to adapt existing XMLRPC handlers to HTTP using Spark.

Authentication is handled by a token passed by a cookie named `pxt-session-cookie`. The cookie with the token can be retrieved via the following request and can be sent in any further request to the API.
```
POST <server>/rhn/manager/api/login
{
  'login': '<username>',
  'password': '<password>'
}
```
**IMPORTANT:** ~~The code needs the latest version of GSON (2.9.0) to run. That's the reason unit tests fail on the PR.~~ The code has been patched to work with GSON 2.8.5.

Best reviewed **commit by commit**!

## To Do
- [x] Insert `@ReadOnly` and `@ApiIgnore` annotations where required
- [x] Test remaining parameter types
- [x] Test all return types
- [x] Deprecate `SnapshotHandler.listSnapshots` and implement an alternative
- [x] Test a read-only user
- [x] Clean up redundant `response.type("application/json")` settings
- [x] Address inline `TODO`s
- [ ] Revert "Support gson-2.8.5" when 2.8.9 is available

### List projects
```
GET <server>/rhn/manager/api/contentmanagement/listProjects
```

### Lookup projects
```
GET <server>/rhn/manager/api/contentmanagement/lookupProject?label=<label>
```

### Create project
```
POST <server>/rhn/manager/api/contentmanagement/createProject
{
  'label': '<label>',
  'name': '<name>',
  'description': '<description>'
}
```

### Create project (alternative)
```
POST <server>/rhn/manager/api/contentmanagement/createProject?label=<label>&name=<name>&description=<desc>
(empty body)
```

### Update project
```
POST <server>/rhn/manager/api/contentmanagement/createProject
{
  'label': '<label>',
  'props': {
    'name': '<name>',
    'description': '<description>'
  }
}
```

### Remove project
```
POST <server>/rhn/manager/api/contentmanagement/removeProject?label=<label>
(empty body)
```

*Parameters of primitives or arrays of primitives may either be passed in the query string or the JSON body, whereas `object` types must always be passed in the JSON body.*

## GUI diff

No difference.

## Documentation
https://github.com/SUSE/spacewalk/issues/17269

## Test coverage
Unit tests are added

## Links

Fixes https://github.com/SUSE/spacewalk/issues/17267

## Changelogs

Make sure the changelogs entries you are adding are compliant with https://github.com/uyuni-project/uyuni/wiki/Contributing#changelogs and https://github.com/uyuni-project/uyuni/wiki/Contributing#uyuni-projectuyuni-repository

If you don't need a changelog check, please mark this checkbox:

- [ ] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_pgsql_tests" 
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

**Comment by Bischoff** [comment]
A small wish:

When calling `api/login` and forgetting `Content-Type: application/json` header, do not return `{"success":true, ...}`, but do return an error instead.

**Comment by Bischoff** [comment]
Another gotcha:

Entry point for login in XML-RPC is `auth.login`, while it's `login` in HTTP. I suggest to harmonize to `auth.login` or at least to provide an alias.

This is to avoid a gotcha when porting scripts from XML-RPC to HTTP, to keep our API documentation unchanged, etc.

**Comment by Bischoff** [comment]
In XML-RPC you could both call `list_users()` and `listUsers`, while in HTTP only `listUsers` works.

Note: maybe it's just a convenience facility from the Ruby XML-RPC client library. It's possible it does not come from SUSE Manager. I don't know.

**Review by parlt91 — APPROVED** [review]
I am going to approve this one now. I spend quite some time looking at it and trying it out, although I'd liked to have some more time (that we don't really have since it has to be merged tomorrow). I did not discover anything that would need bigger adjustments.
It looks really good and I can totally see the effort you put into this including all the tests. That's why I feel relatively confident approving it.
Thank you 😄

File: java/code/src/com/redhat/rhn/frontend/xmlrpc/channel/software/ChannelSoftwareHandler.java [hunk 1]
```diff
@@ -135,6 +136,7 @@ public ChannelSoftwareHandler(TaskomaticApi taskomaticApiIn, XmlRpcSystemHelper
      *
      * @xmlrpc.ignore
```
[the remark this answers is not in the document]
Comment (cbbayburt) [reply java/code/src/com/redhat/rhn/frontend/xmlrpc/channel/software/ChannelSoftwareHandler.java:138]: For now, yes. `@xmlrpc.ignore` excludes the method from the API docs. It won't be needed after https://github.com/SUSE/spacewalk/issues/17269

**Pull request merged**